rust/sbp/src/sbp_tools.rs

```diff
@@ -7,7 +9,47 @@ use crate::{
     time::{GpsTimeError, MessageTime, RoverTime},
 };
 
-use crate::messages::SBP;
+use crate::messages::{MessageType, SBP};
+
+pub struct Dispatcher {
+    callbacks: HashMap<u16, Vec<Box<dyn FnMut(SBP)>>>,
+}
+
+impl Dispatcher {
+    pub fn new() -> Self {
+        Self {
+            callbacks: HashMap::new(),
+        }
+    }
+
+    pub fn on<F, M>(&mut self, mut func: F)
+    where
+        F: FnMut(M) + 'static,
```

Comment:
I'm sure it'll make things much more complicated, but it would be awesome if we could drop the static lifetime here.

Reply:
Hmm not 100% sure how to do it/if it's possible. Afaik boxed closures will always need a static lifetime. It might be possible to not box the closures, which would be great. They'll still need to be behind a pointer of some kind because they are unsized. So something like having the callbacks be `&'a dyn FnMut(SBP)`.  So changing

```rust
self.callbacks
    .entry(M::message_type())
    .or_default()
    .push(Box::new(move |msg: SBP| {
        let msg: M = msg.try_into().unwrap();
        func(msg)
    }));
```

to 

```rust 
self.callbacks
    .entry(M::message_type())
    .or_default()
    .push(&move |msg: SBP| {
        let msg: M = msg.try_into().unwrap();
        func(msg)
    });
```

Would let us drop `'static` in favor of generic lifetime `'a`. But, because we actually are storing a new closure rather than the one that was passed by the users you'll get a `temporary value dropped while borrowed` error. Still might be possible just not immediately obvious how to do it

Reply:
I think the boxed closure can have a shorter lifetime than `'static`, `Dispatcher` just needs to be generic over that lifetime. That'll be a limitation of any callback system like this, any references that the function objects capture will have to outlive the `Dispatch` object. I'm not sure how you could get around that.

Here's a very basic example of how I got the lifetimes to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1e74838c33e509a33d637d4a213cc41

Reply:
Ahh I see, worked like a charm!